refactor Compiler + Generators

georgios-ts opened the pull request:
Trigger Github actions.

clausia committed:
First version of exporting to open qasm

clausia committed:
Optional third party libraries

georgios-ts committed:
generators

georgios-ts committed:
Merge remote-tracking branch 'upstream/devel' into generators

georgios-ts committed:
exclude rotations

georgios-ts committed:
refactor Compiler

kottmanj commented on src/tequila/circuit/_gates_impl.py:
```diff
@@ -255,8 +277,17 @@ def __pow__(self, power, modulo=None):
 
 class PowerGateImpl(ParametrizedGateImpl):
 
```
Would be great to have a small doc string, that just comments on the power an parameter attributes

kottmanj commented on src/tequila/circuit/compiler.py:
```diff
@@ -462,7 +478,7 @@ def compile_controlled_rotation(gate: RotationGateImpl, angles: list = None) ->
 @compiler
 def compile_to_cc(gate) -> QCircuit:
```
If I get it right, then the name should probably be adapted. Maybe "compile_to_single_control" (compile_to_c could be misinterpreted)

kottmanj replied:
with pycharm (I think you are using it?) you can to right-click and then "refactor -> rename" then it changes the name throughout the whole package. The compiler flags need to be adapted manually though.

kottmanj commented on src/tequila/circuit/compiler.py:
```diff
@@ -643,10 +593,20 @@ def compile_power_base(gate):
+    elif gate.name == 'X':
         ### off by global phase of Exp[ pi power /2]
         '''
         if we wanted to do it formally we would use the following
```
Nice. Very clear. Global phase is fine. Maybe add "Global phases can be ignored. For controlled gates the global phase becomes a relative phase so this structure is respected in the compilation of controlled gates." or something like this

kottmanj commented:
Fantastic!

clausia committed:
Import/export OpenQASM code and import/export to pyzx package for ZX-Calculus operations

ajustes

Validate import pyzx and validate multiple targets and max 2 controls

compile_ry() consider all targets, and small fix name

improve compile_y() method

Add flag for zx-calculus and break down controlled rotations not supported by qasm

set default value

Tests for compile_ry and compile_y

Convinient unitary gates: U, u1, u2, u3 and its tests

Import QASM code logic, including custom gates

Pyzx doesn't understand controlled-H, so a compile option is added to support CH for pyzx

clausia committed:
Merge branch 'devel' into add-pyzx-and-qasm

clausia committed:
improve comment

clausia committed:
Better equivalences for CH and Y that includes the power parameter

clausia committed:
fix comments in compiler.py

clausia committed:
dd matrix representation of U gate

clausia committed:
More tests for gates u1,u2,u3,U

clausia committed:
Improve qubits property in QGateImpl

clausia committed:
correct gates equivalent to CH

clausia committed:
Tests for import/export OpenQASM code

clausia committed:
Correct order for U gate equivalent circuit and improvement in tests for u1,u2,u3 gates

clausia committed:
Error (and fix) because changes in _gates_impl.py regarding _qubits and _max_qubit

clausia committed:
Tests for import/export OpenQASM code

clausia committed:
Remove temporal comment

clausia committed:
Avoid some warnings

clausia committed:
Tests for convert from/to pyzx

clausia committed:
Test if a circuit reduced within pyzx remains the same when returned to tequila

clausia committed:
Skip pyzx tests if pyzx package is not installed

clausia committed:
Improve tests for new compile functions (ry, y, ch)

clausia committed:
Merge branch 'devel' into add-pyzx-and-qasm

clausia committed:
Variables are not needed when importing OpqnQASM code

clausia committed:
Merge branch 'devel' into add-pyzx-and-qasm

georgios-ts committed:
minor fixes

georgios-ts committed:
Merge remote-tracking branch 'upstream/devel' into compiler

georgios-ts committed:
fix syntax error

clausia committed:
Merge branch 'merge-with-georgios' into compiler
